File: graph_algorithms/week2_graph_decomposition2/1_cs_curriculum/acyclicity.cpp
#include <iostream>
#include <vector>

using std::pair;
using std::vector;

bool isCyclicDFS(vector<vector<int>> &adj, vector<bool> &visited, vector<bool> &visitedDuringCurrentTraversal, int v) {
  visited[v] = true;
  visitedDuringCurrentTraversal[v] = true;

  for (int w : adj[v])
    if (visitedDuringCurrentTraversal[w] || isCyclicDFS(adj, visited, visitedDuringCurrentTraversal, w))
      return true;

  visitedDuringCurrentTraversal[v] = false;  // Reset before next traversal
  return false;
}

// Returns 0 if the directed graph is acyclic, 1 otherwise.
int acyclic(vector<vector<int>> &adj) {
  int n = adj.size();
  vector<bool> visited(n, false);
  vector<bool> visitedDuringCurrentTraversal(n, false);

  for (int v = 0; v < n; v++)
    if (!visited[v] && isCyclicDFS(adj, visited, visitedDuringCurrentTraversal, v))
      return 1;

  return 0;
}

int main() {
  size_t n, m;
  std::cin >> n >> m;

  vector<vector<int>> adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int a, b;
    std::cin >> a >> b;
    adj[a - 1].push_back(b - 1);
  }

  std::cout << acyclic(adj) << std::endl;
}

File: graph_algorithms/week2_graph_decomposition2/1_cs_curriculum/acyclicity.py
from typing import List


def is_cyclic_dfs(
    adj: List[List[int]],
    visited: List[bool],
    visited_during_current_traversal: List[bool],
    v: int,
) -> bool:
    visited[v] = True
    visited_during_current_traversal[v] = True

    for w in adj[v]:
        if visited_during_current_traversal[w] or is_cyclic_dfs(adj, visited, visited_during_current_traversal, w):
            return True

    visited_during_current_traversal[v] = False  # Reset before next traversal
    return False


def acyclic(adj: List[List[int]]) -> int:
    """Returns 0 if the directed graph is acyclic, 1 otherwise."""
    n = len(adj)
    visited = [False] * n
    visited_during_current_traversal = [False] * n

    for v in range(n):
        if not visited[v] and is_cyclic_dfs(adj, visited, visited_during_current_traversal, v):
            return 1

    return 0


if __name__ == "__main__":
    n, m = map(int, input().split())

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    for a, b in edges:
        adj[a - 1].append(b - 1)

    print(acyclic(adj))
